corehq/apps/saved_reports/models.py

```diff
@@ -127,18 +127,48 @@ def by_domain_and_owner(cls, domain, owner_id, report_slug=None,
         if limit is not None:
             kwargs['limit'] = limit
 
-        result = cache_core.cached_view(
-            db,
-            "reportconfig/configs_by_domain",
-            reduce=False,
-            include_docs=True,
-            startkey=key,
-            endkey=key + [{}],
-            wrapper=cls.wrap,
-            **kwargs
-        )
+        if not include_shared:
+            result = cache_core.cached_view(
+                db,
+                "reportconfig/configs_by_domain",
+                reduce=False,
+                include_docs=True,
+                startkey=key,
+                endkey=key + [{}],
+                wrapper=cls.wrap,
+                **kwargs
+            )
+        else:
+            # Get user's own configs on domain
+            user_configs = cache_core.cached_view(
+                db,
+                "reportconfig/configs_by_domain",
+                reduce=False,
+                include_docs=False,
+                startkey=key,
+                endkey=key + [{}],
+                wrapper=cls.wrap,
+                **kwargs
+            )
+            user_configs = [c['id'] for c in user_configs]
+            # Include all shared configs
+            user_configs.extend(cls.shared_on_domain(domain, only_id=True))
+
+            result = [ReportConfig.get(id) for id in set(user_configs)]
+
```

Comment:
I don't think you need to pull all the configs twice? Also seems like there is a bunch of duplicate code here:

```suggestion
     user_configs = cache_core.cached_view(
                db,
                "reportconfig/configs_by_domain",
                reduce=False,
                include_docs=True,
                startkey=key,
                endkey=key + [{}],
                wrapper=cls.wrap,
                **kwargs
        )
        if include_shared:
            user_config_ids = [c['id'] for c in user_configs]
            shared_configs = [c for c in cls.shared_on_domain(domain) if c._id not in user_config_ids]
            result = user_configs + shared_configs
```

Reply:
This looks much better!